## ДЗ 1 (ОБЯЗАТЕЛЬНОЕ): Анализ температурных данных и мониторинг текущей температуры через OpenWeatherMap API

**Описание задания:**  
Вы аналитик в компании, занимающейся изучением климатических изменений и мониторингом температур в разных городах. Вам нужно провести анализ исторических данных о температуре для выявления сезонных закономерностей и аномалий. Также необходимо подключить API OpenWeatherMap для получения текущей температуры в выбранных городах и сравнить её с историческими данными.


### Цели задания:
1. Провести **анализ временных рядов**, включая:
   - Вычисление скользящего среднего и стандартного отклонения для сглаживания температурных колебаний.
   - Определение аномалий на основе отклонений температуры от $ \text{скользящее среднее} \pm 2\sigma $.
   - Построение долгосрочных трендов изменения температуры.
   - Любые дополнительные исследования будут вам в плюс.

2. Осуществить **мониторинг текущей температуры**:
   - Получить текущую температуру через OpenWeatherMap API.
   - Сравнить её с историческим нормальным диапазоном для текущего сезона.

3. Разработать **интерактивное приложение**:
   - Дать пользователю возможность выбрать город.
   - Отобразить результаты анализа температур, включая временные ряды, сезонные профили и аномалии.
   - Провести анализ текущей температуры в контексте исторических данных.
   - Любые графики должны быть интерактивными.


### Описание данных
Исторические данные о температуре содержатся в файле `temperature_data.csv`, включают:
  - `city`: Название города.
  - `timestamp`: Дата (с шагом в 1 день).
  - `temperature`: Среднесуточная температура (в °C).
  - `season`: Сезон года (зима, весна, лето, осень).

Код для генерации файла вы найдете ниже.

### Этапы выполнения

1. **Анализ исторических данных**:
   - Вычислить **скользящее среднее** температуры с окном в 30 дней для сглаживания краткосрочных колебаний.
   - Рассчитать среднюю температуру и стандартное отклонение для каждого сезона в каждом городе.
   - Выявить аномалии, где температура выходит за пределы $ \text{среднее} \pm 2\sigma $.
   - Попробуйте распараллелить проведение этого анализа. Сравните скорость выполнения анализа с распараллеливанием и без него.

2. **Мониторинг текущей температуры**:
   - Подключить OpenWeatherMap API для получения текущей температуры города. Для получения API Key (бесплатно) надо зарегистрироваться на сайте. Обратите внимание, что API Key может активироваться только через 2-3 часа, это нормально. Посему получите ключ заранее.
   - Получить текущую температуру для выбранного города через OpenWeatherMap API.
   - Определить, является ли текущая температура нормальной, исходя из исторических данных для текущего сезона.
   - Данные на самом деле не совсем реальные (сюрпрайз). Поэтому на момент эксперимента погода в Берлине, Каире и Дубае была в рамках нормы, а в Пекине и Москве аномальная. Протестируйте свое решение для разных городов.
   - Попробуйте для получения текущей температуры использовать синхронные и асинхронные методы. Что здесь лучше использовать?

3. **Создание приложения на Streamlit**:
   - Добавить интерфейс для загрузки файла с историческими данными.
   - Добавить интерфейс для выбора города (из выпадающего списка).
   - Добавить форму для ввода API-ключа OpenWeatherMap. Когда он не введен, данные для текущей погоды не показываются. Если ключ некорректный, выведите на экран ошибку (должно приходить `{"cod":401, "message": "Invalid API key. Please see https://openweathermap.org/faq#error401 for more info."}`).
   - Отобразить:
     - Описательную статистику по историческим данным для города, можно добавить визуализации.
     - Временной ряд температур с выделением аномалий (например, точками другого цвета).
     - Сезонные профили с указанием среднего и стандартного отклонения.
   - Вывести текущую температуру через API и указать, нормальна ли она для сезона.

### Критерии оценивания

- Корректное проведение анализа данных – 1 балл.
- Исследование распараллеливания анализа – 1 балл.
- Корректный поиск аномалий – 1 балл.
- Подключение к API и корректность выполнения запроса – 1 балл.
- Проведение эксперимента с синхронным и асинхронным способом запроса к API – 1 балл.
- Создание интерфейса приложения streamlit в соответствии с описанием – 3 балла.
- Корректное отображение графиков и статистик, а также сезонных профилей – 1 балл.
- Корректный вывод текущей температуры в выбранном городе и проведение проверки на ее аномальность – 1 балл.
- Любая дополнительная функциональность приветствуется и оценивается бонусными баллами (не более 2 в сумме) на усмотрение проверяющего.

### Формат сдачи домашнего задания

Решение нужно развернуть в Streamlit Cloud (бесплатно)

*   Создаем новый репозиторий на GitHub.  
*   Загружаем проект.
*   Создаем аккаунт в [Streamlit Cloud](https://streamlit.io/cloud).
*   Авторизуемся в Streamlit Cloud.
*   Создаем новое приложение в Streamlit Cloud и подключаем GitHub-репозиторий.
*   Deploy!

Сдать в форму необходимо:
1. Ссылку на развернутое в Streamlit Cloud приложение.
2. Ссылку на код. Все выводы про, например, использование параллельности/асинхронности опишите в комментариях.

Не забудьте удалить ключ API и иную чувствительную информацию.

### Полезные ссылки
*   [Оформление задачи Титаник на Streamlit](https://github.com/evgpat/streamlit_demo)
*   [Документация Streamlit](https://docs.streamlit.io/)
*   [Блог о Streamlit](https://blog.streamlit.io/)

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd
import numpy as np

# Реальные средние температуры (примерные данные) для городов по сезонам
seasonal_temperatures = {
    "New York": {"winter": 0, "spring": 10, "summer": 25, "autumn": 15},
    "London": {"winter": 5, "spring": 11, "summer": 18, "autumn": 12},
    "Paris": {"winter": 4, "spring": 12, "summer": 20, "autumn": 13},
    "Tokyo": {"winter": 6, "spring": 15, "summer": 27, "autumn": 18},
    "Moscow": {"winter": -10, "spring": 5, "summer": 18, "autumn": 8},
    "Sydney": {"winter": 12, "spring": 18, "summer": 25, "autumn": 20},
    "Berlin": {"winter": 0, "spring": 10, "summer": 20, "autumn": 11},
    "Beijing": {"winter": -2, "spring": 13, "summer": 27, "autumn": 16},
    "Rio de Janeiro": {"winter": 20, "spring": 25, "summer": 30, "autumn": 25},
    "Dubai": {"winter": 20, "spring": 30, "summer": 40, "autumn": 30},
    "Los Angeles": {"winter": 15, "spring": 18, "summer": 25, "autumn": 20},
    "Singapore": {"winter": 27, "spring": 28, "summer": 28, "autumn": 27},
    "Mumbai": {"winter": 25, "spring": 30, "summer": 35, "autumn": 30},
    "Cairo": {"winter": 15, "spring": 25, "summer": 35, "autumn": 25},
    "Mexico City": {"winter": 12, "spring": 18, "summer": 20, "autumn": 15},
}

# Сопоставление месяцев с сезонами
month_to_season = {12: "winter", 1: "winter", 2: "winter",
                   3: "spring", 4: "spring", 5: "spring",
                   6: "summer", 7: "summer", 8: "summer",
                   9: "autumn", 10: "autumn", 11: "autumn"}

# Генерация данных о температуре
def generate_realistic_temperature_data(cities, num_years=10):
    dates = pd.date_range(start="2010-01-01", periods=365 * num_years, freq="D")
    data = []

    for city in cities:
        for date in dates:
            season = month_to_season[date.month]
            mean_temp = seasonal_temperatures[city][season]
            # Добавляем случайное отклонение
            temperature = np.random.normal(loc=mean_temp, scale=5)
            data.append({"city": city, "timestamp": date, "temperature": temperature})

    df = pd.DataFrame(data)
    df['season'] = df['timestamp'].dt.month.map(lambda x: month_to_season[x])
    return df

# Генерация данных
data = generate_realistic_temperature_data(list(seasonal_temperatures.keys()))
data.to_csv('temperature_data.csv', index=False)


## Анализ исторических данных


In [3]:
import time
from joblib import Parallel, delayed


### 1. Подготовка данных


In [4]:
temp_data = data.copy()
temp_data['timestamp'] = pd.to_datetime(temp_data['timestamp'])
temp_data = temp_data.sort_values(['city', 'timestamp'])
temp_data = temp_data.reset_index(drop=True)
temp_data.head()


,city,timestamp,temperature,season
0,Beijing,2010-01-01,-2.284651,winter
1,Beijing,2010-01-02,0.182642,winter
2,Beijing,2010-01-03,1.438858,winter
3,Beijing,2010-01-04,-1.297941,winter
4,Beijing,2010-01-05,-3.070608,winter


In [5]:
city_grps = []

for city, city_data in temp_data.groupby('city'):
    city_grps.append(city_data)


In [6]:
print(len(city_grps))
city_grps[14].head()

15


,city,timestamp,temperature,season
51100,Tokyo,2010-01-01,13.134876,winter
51101,Tokyo,2010-01-02,-0.778954,winter
51102,Tokyo,2010-01-03,7.546233,winter
51103,Tokyo,2010-01-04,4.828870,winter
51104,Tokyo,2010-01-05,-3.734612,winter


Сгруппировали данные по городам и создали список таблиц по 15 городам, далее это нам понадобится, чтобы считать показатели отдельно для каждого города, а также для параллельной обработки

### 2. Скользящее среднее и стандартное отклонение за 30 дней


In [7]:
def roll_stat_get(city_data):
    res = city_data.sort_values('timestamp').copy()
    res['rolling_mean'] = res['temperature'].rolling(window=30).mean()
    res['rolling_std'] = res['temperature'].rolling(window=30).std()
    return res


In [8]:
roll_parts = []

for city_data in city_grps:
    roll_parts.append(roll_stat_get(city_data))


In [9]:
roll_data = pd.concat(roll_parts, ignore_index=True)


In [10]:
roll_data[['city', 'timestamp', 'temperature', 'rolling_mean', 'rolling_std']].iloc[25:35]


,city,timestamp,temperature,rolling_mean,rolling_std
25,Beijing,2010-01-26,-6.870546,NaN,NaN
26,Beijing,2010-01-27,8.036716,NaN,NaN
27,Beijing,2010-01-28,-8.687433,NaN,NaN
28,Beijing,2010-01-29,0.773615,NaN,NaN
29,Beijing,2010-01-30,-3.253506,-1.220061,5.163622
30,Beijing,2010-01-31,-6.744332,-1.368717,5.258648
31,Beijing,2010-02-01,6.829429,-1.147157,5.462342
32,Beijing,2010-02-02,-3.206810,-1.302013,5.452343
33,Beijing,2010-02-03,1.872217,-1.196341,5.483059
34,Beijing,2010-02-04,-6.093405,-1.297101,5.546101


Для каждого города первые 29 значений скользящего среднего и стандартного отклонения отсутствуют, потому что полного окна из 30 дней еще нет

### 3. Средняя температура и стандартное отклонение по сезонам и городам


In [11]:
def season_stat_get(data):
    res = data.groupby(['city', 'season'])['temperature'].agg(['mean', 'std'])
    res = res.reset_index()
    res = res.rename(columns={'mean': 'season_mean', 'std': 'season_std'})
    return res


In [12]:
season_statistics = season_stat_get(temp_data)


In [13]:
season_statistics.round(2)


,city,season,season_mean,season_std
0,Beijing,autumn,16.01,5.15
1,Beijing,spring,12.93,5.01
2,Beijing,summer,27.06,4.93
3,Beijing,winter,-1.81,4.81
4,Berlin,autumn,10.93,5.10
5,Berlin,spring,9.84,5.03
6,Berlin,summer,20.00,4.94
7,Berlin,winter,-0.09,4.93
8,Cairo,autumn,24.72,4.88
9,Cairo,spring,24.98,5.15


### 4. Аномалии относительно сезонной нормы и скользящего среднего


In [14]:
def season_anom_add(data, stats):
    res = data.merge(stats, on=['city', 'season'])
    res['season_lower'] = res['season_mean'] - 2 * res['season_std']
    res['season_upper'] = res['season_mean'] + 2 * res['season_std']
    res['season_anomaly'] = (
        (res['temperature'] < res['season_lower'])
        | (res['temperature'] > res['season_upper'])
    )
    return res


In [15]:
season_res = season_anom_add(roll_data, season_statistics)


In [16]:
season_anom = season_res[season_res['season_anomaly']]
season_anom[['city', 'timestamp', 'season', 'temperature', 'season_lower', 'season_upper']].head(10)


,city,timestamp,season,temperature,season_lower,season_upper
14,Beijing,2010-01-15,winter,8.860619,-11.424179,7.807754
15,Beijing,2010-01-16,winter,9.683446,-11.424179,7.807754
26,Beijing,2010-01-27,winter,8.036716,-11.424179,7.807754
35,Beijing,2010-02-05,winter,-13.967348,-11.424179,7.807754
69,Beijing,2010-03-11,spring,25.887501,2.908918,22.958948
74,Beijing,2010-03-16,spring,1.548600,2.908918,22.958948
80,Beijing,2010-03-22,spring,24.904639,2.908918,22.958948
183,Beijing,2010-07-03,summer,16.378112,17.206456,36.914497
186,Beijing,2010-07-06,summer,37.018929,17.206456,36.914497
249,Beijing,2010-09-07,autumn,4.890382,5.713917,26.302939


In [17]:
def roll_anom_add(data):
    res = data.copy()
    res['rolling_lower'] = res['rolling_mean'] - 2 * res['rolling_std']
    res['rolling_upper'] = res['rolling_mean'] + 2 * res['rolling_std']
    res['rolling_anomaly'] = (
        (res['temperature'] < res['rolling_lower'])
        | (res['temperature'] > res['rolling_upper'])
    )
    return res


In [18]:
analysis_res = roll_anom_add(season_res)


In [19]:
roll_anomalies = analysis_res[analysis_res['rolling_anomaly']]
roll_anomalies[['city', 'timestamp', 'temperature', 'rolling_lower', 'rolling_upper']].head(10)


,city,timestamp,temperature,rolling_lower,rolling_upper
35,Beijing,2010-02-05,-13.967348,-12.975542,10.144935
59,Beijing,2010-03-01,11.176664,-13.182811,10.393383
60,Beijing,2010-03-02,18.640427,-14.238538,13.141428
62,Beijing,2010-03-04,15.341850,-14.815276,15.068625
69,Beijing,2010-03-11,25.887501,-13.955743,21.964412
132,Beijing,2010-05-13,3.723426,4.205638,17.644688
133,Beijing,2010-05-14,19.324516,3.806627,18.361849
135,Beijing,2010-05-16,21.976439,3.440792,20.201002
151,Beijing,2010-06-01,24.856664,1.917387,23.095770
152,Beijing,2010-06-02,34.235131,0.191070,26.489104


In [20]:
anom_summary = analysis_res.groupby('city')[['season_anomaly', 'rolling_anomaly']].sum()
anom_summary


,season_anomaly,rolling_anomaly
city,,
Beijing,173,234
Berlin,155,192
Cairo,168,221
Dubai,158,198
London,159,162
Los Angeles,152,156
Mexico City,173,155
Moscow,160,242
Mumbai,158,155


Показатели аномалий:

  - **сезонный показатель** 'season_anomaly' - узнаем, насколько температура необычна для конкретного времени года в конкретном городе
  - показатель по **скользящему окну** 'rolling_anomaly' - показывает, вышла ли фактическая температура в конкретный день за границы
  скользящего среднего за 30 дней ± 2 скользящих стандартных отклонения


  Выводы по таблице:

  - Данные меняются от запуска к запуску, обычно сезонных аномалий меньше, чем по скользящему окну
  - Количество сезонных аномалий от 152 до 185 дня. Больше всего аномалий в Токио — 185 дней, а меньше всего в Лос-Анджелесе — 152 дня
  - Количество аномалий относительно скользящего окна от 139 до 242 дней. Больше всего таких аномалий в Москве — 242 дня и Пекине — 234 дня, меньше всего в Сингапуре — 139 дней
  - То есть, в данных Москвы и Пекина температура чаще резко отклонялась от среднего уровня предыдущих 29 дней и текущего, что может указывать на более частые краткосрочные изменения температуры


### 5. Последовательный и параллельный анализ всех городов: сравнение


In [21]:
def analyze_city(city_data):
    res = roll_stat_get(city_data)
    stats = season_stat_get(city_data)
    res = season_anom_add(res, stats)
    res = roll_anom_add(res)
    return res


In [22]:
def analyze_sequent(grps):
    res = []

    for city_data in grps:
        res.append(analyze_city(city_data))

    return pd.concat(res, ignore_index=True)


In [23]:
def analyze_parallel(grps):
    res = Parallel(n_jobs=2)(delayed(analyze_city)(grp) for grp in grps)
    return pd.concat(res, ignore_index=True)


In [24]:
def measure(function, grps):
    start_time = time.time()
    res = function(grps)
    diff = time.time() - start_time
    return res, diff


In [25]:
measurements = []

for repeat in range(10):
    if repeat % 2 == 0:
        sequent_res, sequent_time = measure(analyze_sequent, city_grps)
        parallel_res, parallel_time = measure(analyze_parallel, city_grps)
    else:
        parallel_res, parallel_time = measure(analyze_parallel, city_grps)
        sequent_res, sequent_time = measure(analyze_sequent, city_grps)

    pd.testing.assert_frame_equal(sequent_res, parallel_res)
    measurements.append({
        'Повтор': repeat + 1,
        'Последовательно, с': sequent_time,
        'Параллельно, с': parallel_time,
    })


In [33]:
timing_table = pd.DataFrame(measurements)
timing_table.round(4)


,Повтор,"Последовательно, с","Параллельно, с"
0,1,0.1715,1.2992
1,2,0.1800,0.1829
2,3,0.1545,0.1747
3,4,0.1600,0.1800
4,5,0.1692,0.1916
5,6,0.1698,0.1735
6,7,0.1602,0.1727
7,8,0.1569,0.1932
8,9,0.1631,0.2449
9,10,0.2381,0.3108


In [29]:
pd.testing.assert_frame_equal(analysis_res, sequent_res)
print('Результаты всех видов анализов совпадают')


Результаты всех видов анализов совпадают


In [30]:
sequent_median = timing_table['Последовательно, с'].median()
parallel_median = timing_table['Параллельно, с'].median()
speedup = sequent_median / parallel_median


In [31]:
print(f'Медианное время последовательного анализа: {sequent_median:.4f} с')
print(f'Медианное время параллельного анализа: {parallel_median:.4f} с')
print(f'Отношение последовательного времени к параллельному: {speedup:.2f}')

if speedup > 1:
    print('В этом эксперименте параллельный анализ оказался быстрее.')
else:
    print('В этом эксперименте распараллеливание не дало ускорения.')


Медианное время последовательного анализа: 0.1661 с
Медианное время параллельного анализа: 0.1873 с
Отношение последовательного времени к параллельному: 0.89
В этом эксперименте распараллеливание не дало ускорения.


In [32]:
print(f'Количество строк: {len(analysis_res)}')
print(f"Количество городов: {analysis_res['city'].nunique()}")
print(f"Количество пар город–сезон: {len(season_statistics)}")
print(f"Аномалий по сезонной норме: {analysis_res['season_anomaly'].sum()}")
print(f"Аномалий по скользящей норме: {analysis_res['rolling_anomaly'].sum()}")
print(f"Строк без полного окна 30 дней: {analysis_res['rolling_mean'].isna().sum()}")


Количество строк: 54750
Количество городов: 15
Количество пар город–сезон: 60
Аномалий по сезонной норме: 2456
Аномалий по скользящей норме: 2812
Строк без полного окна 30 дней: 435


- для сравнения последовательного и параллельного анализа использовали медиану, так как она меньше зависит от отдельных слишком быстрых или медленных запусков
- при разных запусках параллельный запуск оказывался то быстрее, то медленнее, чем последовательный, чаще медленнее
- в последнем запуске медианное время последовательного анализа - 0.1661 сек, параллельного анализа - 0.1873 сек
- в этом эксперименте параллельная обработка не дала ускорения, так как отношение меньше 1
- параллельный вариант из 10 повторов ни разу не оказался быстрее, в среднем параллельное выполнение потребовало примерно в 1.13 раза больше времени
- первый параллельный запуск оказался медленнее из-за
дополнительного времени на создание рабочих процессов и передачу им
данных
- для этого небольшого набора данных расходы на организацию параллельной обработки оказались больше, чем полученные преимущества, но на большем наборе данных результат может быть другим

### 6. Долгосрочные тренды изменения температуры

In [34]:
import plotly.graph_objects as go

In [35]:
trend_data = temp_data.copy()
trend_data['year'] = trend_data['timestamp'].dt.year

In [36]:
annual_temp = (
    trend_data
    .groupby(['city', 'year'])['temperature']
    .mean()
    .reset_index()
)

annual_temp = annual_temp.rename(
    columns={'temperature': 'year_mean'}
)

annual_temp.head(15)

,city,year,year_mean
0,Beijing,2010,13.544404
1,Beijing,2011,13.586588
2,Beijing,2012,13.573268
3,Beijing,2013,13.535237
4,Beijing,2014,13.605199
5,Beijing,2015,13.628444
6,Beijing,2016,13.392252
7,Beijing,2017,13.793143
8,Beijing,2018,13.795687
9,Beijing,2019,13.808204


In [37]:
def trend_stat_get(data):
    res = []

    for city, city_data in data.groupby('city'):
        slope, intercept = np.polyfit(
            city_data['year'],
            city_data['year_mean'],
            1
        )

        period = city_data['year'].max() - city_data['year'].min()
        period_change = slope * period

        if slope > 0:
            direction = 'повышение'
        elif slope < 0:
            direction = 'снижение'
        else:
            direction = 'без изменений'

        res.append({
            'city': city,
            'trend_per_year': slope,
            'change_for_period': period_change,
            'direction': direction,
        })

    return pd.DataFrame(res)

In [38]:
trend_statistics = trend_stat_get(annual_temp)

trend_statistics.round(3)

,city,trend_per_year,change_for_period,direction
0,Beijing,0.027,0.247,повышение
1,Berlin,0.012,0.105,повышение
2,Cairo,0.002,0.018,повышение
3,Dubai,0.003,0.029,повышение
4,London,-0.021,-0.192,снижение
5,Los Angeles,-0.052,-0.471,снижение
6,Mexico City,0.030,0.268,повышение
7,Moscow,0.072,0.650,повышение
8,Mumbai,0.033,0.295,повышение
9,New York,-0.002,-0.018,снижение


In [39]:
def trend_plot(sel_city):
    city_data = annual_temp[
        annual_temp['city'] == sel_city
    ].copy()

    slope, intercept = np.polyfit(
        city_data['year'],
        city_data['year_mean'],
        1
    )

    city_data['trend_line'] = (
        slope * city_data['year'] + intercept
    )

    fig = go.Figure()

    fig.add_trace(
        go.Scatter(
            x=city_data['year'],
            y=city_data['year_mean'],
            mode='lines+markers',
            name='Средняя температура'
        )
    )

    fig.add_trace(
        go.Scatter(
            x=city_data['year'],
            y=city_data['trend_line'],
            mode='lines',
            name='Линия тренда'
        )
    )

    fig.update_layout(
        title=f'Долгосрочный тренд температуры: {sel_city}',
        xaxis_title='Год',
        yaxis_title='Средняя температура, °C',
        hovermode='x unified'
    )

    return fig

In [40]:
trend_plot('Moscow')

In [41]:
trend_plot('Tokyo')

In [42]:
trend_plot('Mumbai')

In [43]:
max_coeff = trend_statistics.loc[trend_statistics['trend_per_year'].idxmax()]

min_coeff = trend_statistics.loc[trend_statistics['trend_per_year'].idxmin()]

print(
    'Наибольший положительный коэффициент:',
    max_coeff['city'],
    round(max_coeff['trend_per_year'], 3),
    '°C в год'
)

print(
    'Наименьший коэффициент:',
    min_coeff['city'],
    round(min_coeff['trend_per_year'], 3),
    '°C в год'
)

Наибольший положительный коэффициент: Moscow 0.072 °C в год
Наименьший коэффициент: Singapore -0.071 °C в год


- данные также меняются в зависимости от запуска
- наибольший положительный коэффициент тренда получен для Москвы: около 0.072 °C в год, наибольшее снижение получено для Сингапура: около −0.071 °C в год
- все найденные изменения относительно небольшие, ярко выраженного долгосрочного тренда нет, полученные повышения и снижения объясняются скорее случайными колебаниями

## Мониторинг текущей температуры


### 1. Подключение к OpenWeatherMap через API

In [44]:
%pip install requests aiohttp -q


In [45]:
import asyncio
import aiohttp
import requests
from getpass import getpass


In [46]:
weather_url = 'https://api.openweathermap.org/data/2.5/weather'
api_key = getpass('Введите API-ключ OpenWeatherMap: ')

if not api_key:
    raise ValueError('API-ключ не введен.')


Введите API-ключ OpenWeatherMap: ··········


### 2. Получение текущей температуры (синхронно)

In [47]:
def weather_sync_get(city, api_key):
    params = {
        'q': city,
        'appid': api_key,
        'units': 'metric',
    }
    response = requests.get(weather_url, params=params, timeout=10)
    weather = response.json()

    if response.status_code != 200:
        message = weather.get('message', 'неизвестная ошибка')
        raise ValueError(f'Ошибка API {response.status_code}: {message}')

    local_time = pd.to_datetime(
        weather['dt'] + weather['timezone'], unit='s'
    )

    return {
        'city': city,
        'local_time': local_time,
        'current_temp': weather['main']['temp'],
    }


In [48]:
def weather_now(sel_city):
    weather_sync = weather_sync_get(sel_city, api_key)
    return weather_sync

In [49]:
print(weather_now('Moscow'))
print(weather_now('Tokyo'))
weather_now('Madrid')

{'city': 'Moscow', 'local_time': Timestamp('2026-09-05 18:56:53'), 'current_temp': 16.1}
{'city': 'Tokyo', 'local_time': Timestamp('2026-09-06 01:01:14'), 'current_temp': 22.2}


{'city': 'Madrid',
 'local_time': Timestamp('2026-09-05 18:03:02'),
 'current_temp': 33.5}

### 3. Исторический нормальный диапазон текущего сезона

In [50]:
def season_get(timestamp):
    return month_to_season[timestamp.month]


In [51]:
def current_temp_compare(weather, stats):
    city = weather['city']
    season = season_get(weather['local_time'])

    city_stat = stats[(stats['city'] == city) & (stats['season'] == season)].iloc[0]

    lower = city_stat['season_mean'] - 2 * city_stat['season_std']
    upper = city_stat['season_mean'] + 2 * city_stat['season_std']
    normal = lower <= weather['current_temp'] <= upper

    return {
        'city': city,
        'local_time': weather['local_time'],
        'season': season,
        'current_temp': weather['current_temp'],
        'season_mean': city_stat['season_mean'],
        'season_lower': lower,
        'season_upper': upper,
        'status': 'нормальная' if normal else 'аномальная',
    }


In [52]:
def cities_weather_check(sel_cities):
    res = []

    for city in sel_cities:
        weather = weather_sync_get(city, api_key)
        city_res = current_temp_compare(
            weather,
            season_statistics
        )
        res.append(city_res)

    return pd.DataFrame(res).round(2)

In [53]:
all_cities = [
    'New York',
    'London',
    'Paris',
    'Tokyo',
    'Moscow',
    'Sydney',
    'Berlin',
    'Beijing',
    'Rio de Janeiro',
    'Dubai',
    'Los Angeles',
    'Singapore',
    'Mumbai',
    'Cairo',
    'Mexico City'
]

current_temp_table = cities_weather_check(all_cities)
current_temp_table


,city,local_time,season,current_temp,season_mean,season_lower,season_upper,status
0,New York,2026-09-05 11:51:47,autumn,24.94,14.76,4.74,24.78,аномальная
1,London,2026-09-05 16:53:27,autumn,23.57,11.93,1.43,22.44,аномальная
2,Paris,2026-09-05 17:54:12,autumn,25.11,13.05,3.03,23.07,аномальная
3,Tokyo,2026-09-06 01:01:14,autumn,22.20,18.08,8.35,27.80,нормальная
4,Moscow,2026-09-05 18:55:53,autumn,16.10,7.86,-2.48,18.20,нормальная
5,Sydney,2026-09-06 02:03:17,autumn,18.68,20.25,10.17,30.33,нормальная
6,Berlin,2026-09-05 17:56:26,autumn,16.29,10.93,0.73,21.14,нормальная
7,Beijing,2026-09-06 00:01:41,autumn,23.94,16.01,5.71,26.30,нормальная
8,Rio de Janeiro,2026-09-05 13:03:17,autumn,26.61,24.85,14.69,35.01,нормальная
9,Dubai,2026-09-05 19:57:03,autumn,36.96,29.98,20.01,39.95,нормальная


- температура считается нормальной, если она находится внутри границ исторического сезонного среднего ± 2 стандартных отклонения, значения на самих границах тоже считаются нормальными
- нормальная ли или аномальная температура зависит от времени запуска, предполагаемые аномальные значения по заданию не всегда оказываются аномальными, как сейчас (живем в эпоху изменения климата))
- Нью-Йорк, Лондон и Париж показали аномальную температуру - температуры выше ожидаемого, а в ожидаемых Москве и Пекине аномалий нет

In [54]:
test_cities = ['Berlin', 'Cairo', 'Dubai', 'Beijing', 'Moscow', 'Tokyo', 'Mumbai']


In [55]:
def weather_many_sync_get(cities, api_key):
    res = []

    for city in cities:
        res.append(weather_sync_get(city, api_key))

    return res


In [56]:
start_time = time.time()
sync_weather = weather_many_sync_get(test_cities, api_key)
sync_time = time.time() - start_time

print(f'Время синхронных запросов: {sync_time:.4f} с')


Время синхронных запросов: 1.0481 с


### 4. Получение текущей температуры (асинхронно)

In [57]:
async def weather_async_get(session, city, api_key):
    params = {
        'q': city,
        'appid': api_key,
        'units': 'metric',
    }

    async with session.get(weather_url, params=params) as response:
        weather = await response.json()

        if response.status != 200:
            message = weather.get('message', 'неизвестная ошибка')
            raise ValueError(f'Ошибка API {response.status}: {message}')

        local_time = pd.to_datetime(
            weather['dt'] + weather['timezone'], unit='s'
        )

        return {
            'city': city,
            'local_time': local_time,
            'current_temp': weather['main']['temp'],
        }


In [58]:
async def weather_many_async_get(cities, api_key):
    timeout = aiohttp.ClientTimeout(total=10)

    async with aiohttp.ClientSession(timeout=timeout) as session:
        tasks = [
            weather_async_get(session, city, api_key)
            for city in cities
        ]
        return await asyncio.gather(*tasks)


In [59]:
start_time = time.time()
async_weather = await weather_many_async_get(test_cities, api_key)
async_time = time.time() - start_time

print(f'Время асинхронных запросов: {async_time:.4f} с')


Время асинхронных запросов: 0.1639 с


In [60]:
async_current_res = []

for weather in async_weather:
    async_current_res.append(
        current_temp_compare(weather, season_statistics)
    )

current_temp_table = pd.DataFrame(async_current_res)
current_temp_table.round(2)


,city,local_time,season,current_temp,season_mean,season_lower,season_upper,status
0,Berlin,2026-09-05 17:57:05,autumn,16.29,10.93,0.73,21.14,нормальная
1,Cairo,2026-09-05 18:55:24,autumn,33.42,24.72,14.97,34.47,нормальная
2,Dubai,2026-09-05 19:58:45,autumn,36.96,29.98,20.01,39.95,нормальная
3,Beijing,2026-09-05 23:56:05,autumn,23.94,16.01,5.71,26.30,нормальная
4,Moscow,2026-09-05 18:56:53,autumn,16.10,7.86,-2.48,18.20,нормальная
5,Tokyo,2026-09-06 00:55:36,autumn,22.20,18.08,8.35,27.80,нормальная
6,Mumbai,2026-09-05 21:33:18,autumn,25.99,30.24,20.57,39.91,нормальная


Результат зависит от фактической погоды в момент запуска, поэтому города, указанные в условии как нормальные или аномальные, могут получить другой статус при новом запуске, в этом запуске не было аномальных температур

### 5. Сравнение

In [61]:
request_timing = pd.DataFrame({
    'Способ': ['Синхронный', 'Асинхронный'],
    'Время, с': [sync_time, async_time],
})
request_timing.round(4)


,Способ,"Время, с"
0,Синхронный,1.0481
1,Асинхронный,0.1639


In [62]:
if async_time < sync_time:
    request_speedup = sync_time / async_time
    print(
        f'Асинхронные запросы оказались быстрее в {request_speedup:.2f} раза.'
    )
else:
    print('В этом запуске асинхронный способ не дал ускорения.')


Асинхронные запросы оказались быстрее в 6.39 раза.


Выводы

- Для одного города удобнее синхронный способ, так как его код проще, а от асинхронности почти нет преимуществ
- Для нескольких городов лучше подходит асинхронный способ: запросы выполняются одновременно во время ожидания ответов от API
- Синхронное получение погоды для семи городов заняло 1.0481 секунды, асинхронное — 0.1639 секунды, в этом запуске асинхронный способ оказался примерно в 6.39 раза быстрее

